gems/aws-sdk-core/lib/aws-sdk-core/middleware/checksums.rb

```diff
@@ -0,0 +1,279 @@
+# frozen_string_literal: true
+
+module AWS::SDK::Core
+  module Middleware
+    # Middleware for computing and verifying checksums for the
+    # aws.protocols#httpChecksum trait
+    # @api private
+    class Checksum
+      # This error indicates a checksum failed.
+      class ChecksumError < RuntimeError; end
+
+      CHUNK_SIZE = 1 * 1024 * 1024 # one MB
+
+      # determine the set of supported client side checksum algorithms
+      # CRC32c requires aws-crt (optional sdk dependency) for support
+      CLIENT_ALGORITHMS = begin
+        supported = %w[SHA256 SHA1 CRC32]
+        supported << 'CRC32C' if AWS::SDK::Core.crt_loaded?
+        supported
+      end.freeze
+
+      # priority order of checksum algorithms to validate responses against
+      # Remove any algorithms not supported by client
+      # (ie, depending on CRT availability)
+      CHECKSUM_ALGORITHM_PRIORITIES =
+        %w[CRC32C SHA1 CRC32 SHA256] & CLIENT_ALGORITHMS
+
+      def initialize(app, request_algorithm_member: nil,
+                     request_checksum_required: false,
+                     streaming: false,
+                     signed_streaming: false,
+                     require_decoded_content_length: false,
+                     request_validation_mode_member: nil,
+                     response_algorithms: nil)
+        @app = app
+        @request_algorithm_member = request_algorithm_member
+        @request_checksum_required = request_checksum_required
+        @request_validation_mode_member = request_validation_mode_member
+        @response_algorithms = response_algorithms
+
+        # Streaming configuration - these could all be a hash instead?
+        @streaming = streaming
+        @signed_streaming = signed_streaming
+        @require_decoded_content_length = require_decoded_content_length
+      end
+
+      def call(input, context)
```

Comment:
I'm assuming you took this mostly from v3? I would prefer if we could clean this one up at some point, like breaking it down further into more methods.

Reply:
No - I took it from the SRA doc as much as possible.

Only a little code is from V3.  but yeah, I can do some cleaning up and breaking down - I mostly wanted it to match the doc as closely as possible while implementing the full functionality.